Common/IO/FileFinder.cs

```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Betlln.IO
+{
+    public class FileFinder : IFileFinder
+    {
+        private readonly IFileSystem _fileSystem;
+
+        public FileFinder(IFileSystem fileSystem)
+        {
+            _fileSystem = fileSystem;
+        }
+
+        public event EventHandler<ProgressChangedEventArgs> ProgressUpdated;
+
+        public string FindFile(FileDemand demand, string folderPath, TimeSpan timeout)
+        {
+            Stopwatch stopwatch = new Stopwatch();
+            stopwatch.Start();
+
+            string downloadedFileName = null;
+            while (stopwatch.Elapsed < timeout && string.IsNullOrWhiteSpace(downloadedFileName))
+            {
+                UpdateProgress(0, "Waiting for file download...");
+                downloadedFileName = FindFile(demand, folderPath);
+            }
+            stopwatch.Stop();
+            UpdateProgress(100, null);
+
+            if (string.IsNullOrWhiteSpace(downloadedFileName))
+            {
+                throw new TimeoutException("The download took too long or did not complete.");
+            }
+
+            return downloadedFileName;
+        }
+
+        private void UpdateProgress(int percentComplete, string message)
+        {
+            ProgressUpdated?.Invoke(this, new ProgressChangedEventArgs(percentComplete, message));
+        }
+
+        public bool CanFindAllFiles(MultiFileDemand demand, string folderPath)
```

Comment:
I know you have seen comments from me before on style, this function here just seems like its doing two things and/or returning in so many weird ways. I'm just commenting to give pause on review. Ignore otherwise.

Reply:
Thanks for the reminder! I found a clearer way to express this.